### ISSUE TITLE
[VarExporter] Doctrine Query Cache Fails Rehydration 

### ISSUE THREAD

**Issue by kiler129** [symfony/symfony#29406]
**Symfony version(s) affected**: 4.2.0

**Description**  
Everything started with the following stacktrace (comming from 4.2.0 + Doctrine ORM v2.6.3):
```
09:40:16 WARNING   [cache] Failed to fetch key "%5Bb72c025cf288ac4aa4ccc822c726a95d%5D%5B1%5D"
[
  "key" => "%5Bb72c025cf288ac4aa4ccc822c726a95d%5D%5B1%5D",
  "exception" => Symfony\Component\VarExporter\Exception\NotInstantiableTypeException {
    #message: "Type "Doctrine\ORM\Query\Exec\AbstractSqlExecutor" is not instantiable."
    #code: 0
    #file: "./vendor/symfony/var-exporter/Internal/Registry.php"
    #line: 76
    trace: {
      ./vendor/symfony/var-exporter/Internal/Registry.php:76 { …}
      ./vendor/symfony/var-exporter/Internal/Hydrator.php:62 { …}
      ./vendor/symfony/var-exporter/Internal/Hydrator.php:41 { …}
      ./var/cache/prod/pools/yrz6rpPlAX/W/1/wZ+dYy9L2pl6hCB0KCTg:140 {
        ›     $o[0],
        ›     []
        › );
        arguments: {
          $objects: [ …3]
          $values: null
          $properties: [ …3]
          $value: Doctrine\ORM\Query\ParserResult { …}
          $wakeups: []
        }
      }
      ./vendor/symfony/cache/Traits/PhpFilesTrait.php:92 { …}
      ./vendor/symfony/cache/Adapter/AbstractAdapter.php:164 { …}
      ./vendor/symfony/cache/DoctrineProvider.php:54 { …}
      ./vendor/doctrine/cache/lib/Doctrine/Common/Cache/CacheProvider.php:60 { …}
      ./vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:271 { …}
      ./vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:231 { …}
      ./vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:902 { …}
      ./vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:667 { …}
      ./src/Command/Import/..../....php:271 { …}
      ./src/Command/Import/..../......:168 { …}
      ./vendor/symfony/console/Command/Command.php:255 { …}
      ./vendor/symfony/console/Application.php:919 { …}
      ./vendor/symfony/framework-bundle/Console/Application.php:89 { …}
      ./vendor/symfony/console/Application.php:262 { …}
      ./vendor/symfony/framework-bundle/Console/Application.php:75 { …}
      ./vendor/symfony/console/Application.php:145 { …}
      ./bin/console:44 { …}
    }
  }

```

**How to reproduce**  
The code causing it is basically this:
```
$qb = $this->providerRepo->createQueryBuilder('
$q = $qb->getQuery();
$qi = $q->iterate();
```

First time it runs the cache file is saved: https://pastebin.com/raw/F0L5GKst
Next time code runs the cache file cannot be hydrated and causes the exception mentioned.

As @greg0ire suggested on Slack I dumped the object just [before it's saved to cache](https://github.com/doctrine/doctrine2/blob/v2.6.3/lib/Doctrine/ORM/Query.php#L285): https://pastebin.com/raw/pn3eQUAC

**Possible Solution**  
Unfortunately I have no idea how to fix the problem.

//ping @nicolas-grekas

**Comment by yarhon** [comment]
I was in the process of adding details and unit test for this bug (I've came to it via the symfony/cache),
when seen the new patch.

Not 100% sure in my conclusions (VarExporter logic isn't very simple),
but it seems to me this bug's root reason is not the property(s) access modifier, but the parent abstract class.

`Internal\Exporter::prepare` processes properties of the parent abstract class and links them to that abstract class, then `Internal\Hydrator::hydrate` tries to hydrate this property(s), calling `Internal\Registry::getClassReflector`, which fails because abstract class is not instantiable.

P.S. amazed with the reaction time.

**Comment by yarhon** [comment]
For reference, my unit test

```
use PHPUnit\Framework\TestCase;
use Symfony\Component\VarExporter\VarExporter;

abstract class AbstractItem
{
    private $property;

    public function __construct($property)
    {
        $this->property = $property;
    }
}

class Item extends AbstractItem
{
}

class VarExporterTest extends TestCase
{
    public function testExport()
    {
        $value = new Item('foo');
        $isStaticValue = true;

        $valueExported = VarExporter::export($value, $isStaticValue);
        $valueHydrated = eval('return '.$valueExported.';');

        $this->assertEquals($value, $valueHydrated);
    }
}
```

**Comment by nicolas-grekas** [comment]
Thanks for investigating @yarhon

**Comment by yarhon** [comment]
I'm sure the bug is still there. I've extended the previous test and just tried to run the following test on the latest 4.2 branch - VarExporter throws exceptions when trying hydrate object of class that has abstract parent with private properties.
[Unit test](https://pastebin.com/jGwHSUAs)

**Comment by nicolas-grekas** [comment]
@yarhon thanks, fixed in #29446
